kernel/arch/dreamcast/hardware/g2dma.c

```diff
@@ -0,0 +1,234 @@
+    int chn = code - ASIC_EVT_G2_DMA0;
+
+    if(chn < 0 || chn > 3) {
+        dbglog(DBG_INFO, "g2_dma: Wrong channel received in g2_dma_irq_hnd");
+        chn = 1;
+    }
+
+    /* VP : changed the order of things so that we can chain dma calls */
+
+    /* Signal the calling thread to continue, if any. */
+    if(dma_blocking[chn]) {
+        sem_signal(&dma_done[chn]);
+        thd_schedule(1, 0);
+        dma_blocking[chn] = 0;
+    }
+
+    /* Call the callback, if any. */
+    if(dma_callback[chn]) {
+        dma_callback[chn](dma_cbdata[chn]);
+    }
+}
+
+int g2_dma_transfer(void * sh4, void * g2bus, uint32_t length, uint32_t block,
+                    g2_dma_callback_t callback, ptr_t cbdata,
+                    uint32_t dir, uint32_t mode, uint32_t g2chn, uint32_t sh4chn) {
+    /* No longer used but we keep then around for compatibility */
+    (void)mode;
+    (void)sh4chn;
+
+    if(g2chn > 3) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    /* Make sure we're not already DMA'ing */
+    if(extdma->dma[g2chn].start != 0) {
+        dbglog(DBG_ERROR, "g2_dma: Already DMA'ing for channel %ld\n", g2chn);
+        errno = EINPROGRESS;
+        return -1;
+    }
+
+    /* Check alignments */
+    if(((uint32_t)sh4) & 31) {
+        dbglog(DBG_ERROR, "g2_dma: Unaligned sh4 DMA %p\n", sh4);
+        errno = EFAULT;
+        return -1;
+    }
+
+    if(((uint32_t)g2bus) & 31) {
+        dbglog(DBG_ERROR, "g2_dma: Unaligned g2bus DMA %p\n", g2bus);
+        errno = EFAULT;
+        return -1;
+    }
+
+    /* Make sure length is a multiple of 32 */
+    length = (length + 0x1f) & ~0x1f;
+
+    dma_blocking[g2chn] = block;
+    dma_callback[g2chn] = callback;
+    dma_cbdata[g2chn] = cbdata;
+
+    /* End current DMA transfer */
+    dma_disable(g2chn);
+
+    /* Set needed registers */
+    extdma->dma[g2chn].ext_addr = ((uint32_t)g2bus) & MASK_ADDRESS;
+    extdma->dma[g2chn].sh4_addr = ((uint32_t)sh4) & MASK_ADDRESS;
+    extdma->dma[g2chn].size = length | RESET_ENABLED;
+    extdma->dma[g2chn].dir = dir;
+    extdma->dma[g2chn].trigger_select = CPU_TRIGGER;
+
+    /* Start the DMA transfer */
+    extdma->dma[g2chn].enable = 1;
+    extdma->dma[g2chn].start = 1;
+
+    /* Wait for us to be signaled */
+    if(block)
+        sem_wait(&dma_done[g2chn]);
+
+    return 0;
+}
+
+int g2_dma_init(void) {
+    int i;
+
+    if(dma_init)
+        return 0;
+
+    dma_init = 1;
+
+    for(i = 0; i < 4; i++) {
+        /* Create an initially blocked semaphore */
+        sem_init(&dma_done[i], 0);
+        dma_blocking[i] = 0;
+        dma_callback[i] = NULL;
+        dma_cbdata[i] = 0;
+
+        /* Hook the interrupt */
+        asic_evt_set_handler(ASIC_EVT_G2_DMA0 + i, g2_dma_irq_hnd);
+        asic_evt_enable(ASIC_EVT_G2_DMA0 + i, ASIC_IRQ);
+    }
+
+    /* Setup the DMA transfer on the external side */
+    extdma->ds_timeout = DS_CYCLE_OVERRIDE;
+    extdma->protection = DISABLE_SYS_MEM_PROTECTION; 
```

Comment:
Should this be re-set to enable on dma_shutdown? Additionally, based on the description of this, shouldn't/couldn't this be set per-transfer to reflect the scope of memory being exempted from protection?

Reply:
The g2 dma is only initialized and shutdown (closing the app)once.  Yes!  It can be set per transfer but I kept it in its original spot.  I dont see it being updated often or at all so I think we can set it just once.